File: main_project/tripod-new-branch/src/java/TripodConfig/src/main/java/com/renren/tripod/configclient/ProducerManagerConfigClient.java
package com.renren.tripod.configclient;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

import com.renren.tripod.ProducerManagerPrx;
import com.renren.tripod.ProducerManagerPrxHelper;
import com.renren.tripod.configutil.ConfigClient;
import com.renren.tripod.configutil.ConfigData;

public class ProducerManagerConfigClient extends ConfigClient {
	private static org.apache.commons.logging.Log _logger = org.apache.commons.logging.LogFactory
			.getLog(ProducerManagerConfigClient.class);
	private Ice.Communicator _ic = Ice.Util.initialize();
	private HashHolder<ProducerManagerPrx> _ccHashHolder = new HashHolder<ProducerManagerPrx>(
			100);
	private HashHolder<ProducerManagerPrx> _javaHashHolder = new HashHolder<ProducerManagerPrx>(
			100);

	public ProducerManagerConfigClient(String zkAddress, String identifier) {
		super(zkAddress, identifier);
		_logger.info("ProducerManagerConfigClient.ProducerManagerConfigClient zkAddress "
				+ zkAddress + " identifier " + identifier);
	}

	@Override
	protected boolean update(ConfigData newConfig) {
		_logger.info("ProducerManagerConfigClient.update");
		ConfigData cmConfig = newConfig._childrenDatas
				.get(ConfigData.TRIPOD_MANAGER);
		if (cmConfig == null) {
			_logger.error("ProducerManagerConfigClient.update cmConfig == null!");
			return false;
		}

		ConfigData ccConfig = cmConfig._childrenDatas
				.get(ConfigData.TRIPOD_PRODUCER_CC);
		if (ccConfig == null) {
			_logger.error("ProducerManagerConfigClient.update ccConfig == null!");
			return false;
		}

		ConfigData javaConfig = cmConfig._childrenDatas
				.get(ConfigData.TRIPOD_PRODUCER_JAVA);
		if (javaConfig == null) {
			_logger.error("ProducerManagerConfigClient.update javaConfig == null!");
			return false;
		}

		Map<String, ProducerManagerPrx> ccClients = new HashMap<String, ProducerManagerPrx>();
		for (Entry<String, ConfigData> ent : ccConfig._childrenDatas.entrySet()) {
			String[] parts = ent.getKey().split(":");
			if (parts.length != 2) {
				return false;
			}
			ProducerManagerPrx prx = ProducerManagerPrxHelper.uncheckedCast(_ic
					.stringToProxy("M:tcp -h " + parts[0] + " -p " + parts[1] + " -t " + 1000));
			if (prx == null) {
				_logger.error("ProducerManagerConfigClient.update prx == null for string "
						+ ent.getKey());
				return false;
			}
			ccClients.put(ent.getKey(), prx);
		}
		_ccHashHolder.renew(ccClients);

		Map<String, ProducerManagerPrx> javaClients = new HashMap<String, ProducerManagerPrx>();
		for (Entry<String, ConfigData> ent : javaConfig._childrenDatas
				.entrySet()) {
			String[] parts = ent.getKey().split(":");
			if (parts.length != 2) {
				return false;
			}
			ProducerManagerPrx prx = ProducerManagerPrxHelper.uncheckedCast(_ic
					.stringToProxy("M:tcp -h " + parts[0] + " -p " + parts[1] + " -t " + 1000));
			if (prx == null) {
				_logger.error("ProducerManagerConfigClient.update prx == null for string "
						+ ent.getKey());
				return false;
			}
			javaClients.put(ent.getKey(), prx);
		}
		_javaHashHolder.renew(javaClients);
		return true;
	}

	public ProducerManagerPrx getCCPrx(String key) {
		return _ccHashHolder.get(key);
	}

	public ProducerManagerPrx getJavaPrx(String key) {
		return _javaHashHolder.get(key);
	}

	public Map<ProducerManagerPrx, List<String>> getCCPrxToKeysMap(
			List<String> keys) {
		return _ccHashHolder.get(keys);
	}

	public Map<ProducerManagerPrx, List<String>> getJavaPrxToKeysMap(
			List<String> keys) {
		return _javaHashHolder.get(keys);
	}

}

File: main_project/tripod-new-branch/src/cpp/configclient/ProducerManagerConfigClient.cpp
#include <boost/algorithm/string/split.hpp>
#include <boost/algorithm/string/classification.hpp>
#include <Common.h>
#include <ConfigData.h>
#include <ProducerManagerConfigClient.h>

namespace com {
namespace renren {
namespace tripod {

ProducerManagerConfigClient::ProducerManagerConfigClient(const std::string& zkAddress) : ConfigClient(zkAddress, TRIPOD_PRODUCER) {
    MCE_INFO("ProducerManagerConfigClient::ProducerManagerConfigClient zkAddress " << zkAddress);
    ic_ = Ice::initialize();
}

bool ProducerManagerConfigClient::update(const ConfigDataPtr& newConfig) {
    std::map<std::string, ConfigDataPtr>::const_iterator findIt = newConfig->childrenDatas_.find(TRIPOD_PRODUCER);
    if(findIt == newConfig->childrenDatas_.end()) {
        MCE_WARN("ProducerManagerConfigClient::update no produceremanager in config!");
        return false;
    }

    ConfigDataPtr producerConfig = findIt->second;

    findIt = producerConfig->childrenDatas_.find(TRIPOD_PRODUCER_JAVA);
    if(findIt == producerConfig->childrenDatas_.end()) {
        MCE_WARN("ProducerManagerConfigClient::update no java produceremanager in config!");
        return false;
    }

    std::map<std::string, ProducerManagerPrx> javaPrxs;
    for(std::map<std::string, ConfigDataPtr>::const_iterator it = findIt->second->childrenDatas_.begin();
            it != findIt->second->childrenDatas_.end(); ++it) {
        std::vector<std::string> parts;
        boost::algorithm::split(parts, it->second->name_, boost::algorithm::is_any_of(":"));
        if(parts.size() != 2) {
            std::ostringstream oss;
            oss << "ProducerManagerConfigClient::update Exception the config string" << it->second->name_
                << " not match the format: <host>:<port>";
            MCE_WARN(oss.str());
            return false;
        }
        std::string address = parts[0];
        std::string port = parts[1];
        std::ostringstream os;
        os<<"M:tcp -h "<<address<<" -p "<<port<<" -t 1000";
        javaPrxs.insert(make_pair(it->second->name_, ProducerManagerPrx::uncheckedCast(ic_->stringToProxy(os.str()))));
    }

    javaPrxHashHolder_.renew(javaPrxs);

    findIt = producerConfig->childrenDatas_.find(TRIPOD_PRODUCER_CC);
    if(findIt == producerConfig->childrenDatas_.end()) {
        MCE_WARN("ProducerManagerConfigClient::update no cc produceremanager in config!");
        return false;
    }

    std::map<std::string, ProducerManagerPrx> ccPrxs;
    for(std::map<std::string, ConfigDataPtr>::const_iterator it = findIt->second->childrenDatas_.begin();
            it != findIt->second->childrenDatas_.end(); ++it) {
        std::vector<std::string> parts;
        boost::algorithm::split(parts, it->second->name_, boost::algorithm::is_any_of(":"));
        if(parts.size() != 2) {
            std::ostringstream oss;
            oss << "ProducerManagerConfigClient::update Exception the config string" << it->second->name_
                << " not match the format: <host>:<port>";
            MCE_WARN(oss.str());
            return false;
        }
        std::string address = parts[0];
        std::string port = parts[1];
        std::ostringstream os;
        os<<"M:tcp -h "<<address<<" -p "<<port<<" -t 1000";
        ccPrxs.insert(make_pair(it->second->name_, ProducerManagerPrx::uncheckedCast(ic_->stringToProxy(os.str()))));
    }

    ccPrxHashHolder_.renew(ccPrxs);

    return true;
}

std::map<ProducerManagerPrx, KeySeq> ProducerManagerConfigClient::getJavaPrxToKeysMap(const KeySeq& keys) {

    return javaPrxHashHolder_.get(keys);
}

std::map<ProducerManagerPrx, KeySeq> ProducerManagerConfigClient::getCcPrxToKeysMap(const KeySeq& keys) {

    return ccPrxHashHolder_.get(keys);
}

std::map<std::string, ProducerManagerPrx> ProducerManagerConfigClient::getOriginalPrxMap() {
    std::map<std::string, ProducerManagerPrx> res;
    std::map<std::string, ProducerManagerPrx> ccRes = ccPrxHashHolder_.getOriginalPrxMap();
    std::map<std::string, ProducerManagerPrx> javaRes = javaPrxHashHolder_.getOriginalPrxMap();
    res.insert(ccRes.begin(), ccRes.end());
    res.insert(javaRes.begin(), javaRes.end());
    return res;
}

}
}
}
